Fix Leaflet.Draw.DrawEvents.Created.layer type

jambonrose opened the pull request:
Editing an existing type definition to fix a type.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://leaflet.github.io/Leaflet.draw/docs/leaflet-draw-latest.html#l-draw-event-draw:created

The `DrawEvents.Created` layer variable is not of type `Layer`. [As noted by the in-code documentation, the `Leaflet.Draw.DrawEvents.Created.layer` variable is one of Leaflet's vector shapes or a markers.](https://github.com/Leaflet/Leaflet.draw/blob/c6af99b761c5d9900bf5a3e2e1d8827bbde29542/src/Leaflet.Draw.Event.js#L31)

[The `L.Draw.Event.CREATED` event is only fired by `Draw.Feature` in the `_fireCreatedEvent` method.](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Feature.js#L94)

Draw.Feature bequeaths _fireCreatedEvent to it's subclasses Draw.Marker, Draw.Polyline, and `Draw.SimpleShape`, which in turn bequeath the method to  `Draw.Circle`, `Draw.Rectangle` (both from `Draw.SimpleShape`), `Draw.CircleMarker` (from `Draw.Marker`), and `Draw.Polygon` (from `Draw.Polyline`).

[`Draw.Circle`](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Circle.js#L56), [`Draw.CircleMarker`](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.CircleMarker.js#L36), [`Draw.Marker`](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Marker.js#L125), [`Draw.Rectangle`](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Rectangle.js#L71) and [`Draw.Polyline`](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Polyline.js#L590) override the `_fireCreatedEvent` to pass instances of their respective vector classes to `_fireCreatedEvent`, meaning that the layer variable is not of type `Layer`, but instead one of the vector or marker classes. For example, `Draw.Circle` passes `L.Circle` to `L.Draw.Feature.prototype._fireCreatedEvent` as the argument to the `layer` parameter.

`Draw.Polygon` inherits `Draw.Polyline's` behavior without changes, but changes the vector class via the [Poly class variable](https://github.com/Leaflet/Leaflet.draw/blob/develop/src/draw/handler/Draw.Polygon.js#L11).

Note that some of the documentation indicates that the `layer` variable may also be a `string`, whereas other parts of the documentation omit this type. I have not seen evidence of this in the code, so I have omitted the `string` type from the new list. Apologies to future users if I missed it.

reblace commented:
Feel free to correct me if I'm wrong, but in each of the handler classes, the ```_fireCreatedEvent``` functions each create a vector feature (e.g., L.Circle, L.Polyline, L.Rectangle) and pass that through the created event. Those vector features each extent L.Layer, so the type definition is correct.

Example:
```
_fireCreatedEvent: function () {
	var circle = new L.Circle(this._startLatLng, this._shape.getRadius(), this.options.shapeOptions);
	L.Draw.SimpleShape.prototype._fireCreatedEvent.call(this, circle);
},
```

```
	Poly: L.Polyline,
	...
	_fireCreatedEvent: function () {
		var poly = new this.Poly(this._poly.getLatLngs(), this.options.shapeOptions);
		L.Draw.Feature.prototype._fireCreatedEvent.call(this, poly);
	}
```

With your change, the type definition is saying that the created event is actually returning a reference to the Leaflet-Draw Feature (e.g., L.Draw.Circle, L.Draw.Polyline, etc). And, I don't think that's actually correct.

jambonrose commented:
Hi @reblace - thanks for the quick response.

You are correct, and I apologize if my original message was confusing: I am trying to change the type of `layer` to reference `L.Circle`, `L.Polyline`, `L.Rectangle`, etc., and not `L.Draw.Circle`, `L.Draw.Polyline`, etc. However, my understanding is that the current code does so. 

I initially specified the following:

```ts
layer: L.Circle | L.CircleMarker | L.Marker | L.Polygon | L.Polyline | L.Rectangle;
```

However, running `npm run lint leaflet-draw` resulted in a number of errors that read `no-unnecessary-qualifier  Qualifier is unnecessary since 'L' is in scope.`. These errors lead me to write the current code.

I wrote the test with the `L` qualifier in place, and I note that running `tsc` in the directory with and without the `L` qualifier works as expected.

To contrast, I replaced the `layer` variable with the following definition.

```ts
layer: Draw.Circle | Draw.CircleMarker | Draw.Marker | Draw.Polygon | Draw.Polyline | Draw.Rectangle;
```

Running `tsc` in the directory results in the following errors (the first is the test I added, whereas the second is an existing test).

```
leaflet-draw-tests.ts(44,29): error TS2339: Property 'toGeoJSON' does not exist on type 'Marker | CircleMarker | Circle | Polyline | Rectangle | Polygon'.
  Property 'toGeoJSON' does not exist on type 'Marker'.
leaflet-draw-tests.ts(46,25): error TS2345: Argument of type 'Marker | CircleMarker | Circle | Polyline | Rectangle | Polygon' is not assignable to parameter of type 'Layer'.
  Type 'Marker' is not assignable to type 'Layer'.
    Property 'addTo' is missing in type 'Marker'.
```

This is to be expected, as `L.Circle` and others have a `toGeoJSON()` method, whereas `L.Draw.Circle` and others do not. As such, my current understanding is that the current change works as desired (but hey, I'm new to Typescript, so a double-check goes a long way).

I concede my initial PR text should be better stated: while I understand that the current `Layer` type is technically correct, I would argue that it is overly broad. By being more precise, we:

- help developers new to leaflet and leaflet-draw (like me!) learn the API
- help developers avoid erroneous type assertions (e.g.: it is possible to assert that `Layer` subclass `GridLayer` will be assigned to `layer`, but that will never actually occur)
- help developers avoid unnecessary type assertions (e.g.: the new test fails with `Layer`, but works with the new definition)

I hope you'll take this developer experience into consideration when reviewing this PR.

reblace commented:
Ahh, I see. Thanks @jambonrose for the clarification. I see what your intention was now. I mistakenly assumed you were trying to type them as the interfaces in the Draw.* namespace.

I don't have an issue with this change. It is more restrictive, which is probably a net benefit (as you state) to users of the library. The added benefit is that it might actually expose more methods/properties on the vector features without requiring a cast to a specific type.
